shiny/_main_create.py

```diff
@@ -37,40 +38,173 @@
+
+    Attributes
+    ----------
+    name
+        The name of the Shiny template.
+    path
+        The path to the `_template.json` file or the root directory of the template.
+    type
+        The type of the template (e.g. "app", "package"), default: 'app'.
+    title
+        A brief title for the template, if provided.
+    description
+        A longer description of the template, if provided.
+    next_steps
+        A list of next steps or instructions related to this template, shown after the
+        default instructions are displayed. In the `_template.json` file, this field
+        can be a single string or an array of strings.
+    follow_up
+        A list of follow-up actions or information related to this template. In the
+        `_template.json` file, this field can be a single string, an array of strings,
+        or an object with a `text` field and an optional `type` field. The `type` field
+        can be "action", "info", "warning", "danger", or "text".
+    """
+
+    name: str
+    path: Path
+    type: str = "app"
+    title: str | None = None
+    description: str | None = None
+    next_steps: list[str] = field(default_factory=list)
+    follow_up: list[ShinyTemplateFollowUp] = field(default_factory=list)
+    _express_available: bool | None = None
+
+    @property
+    def express_available(self) -> bool:
+        """
+        Does the template include an Express variant, denoted by the presence of an
+        `app-express.py` file?
+        """
+
+        if self._express_available is None:
+            self._express_available = (self.path / "app-express.py").exists()
+        return self._express_available
+
+
+class ShinyTemplateFollowUp:
+    def __init__(
+        self,
+        text: str,
+        type: str = "text",
+    ):
+        self.text = text
+        self.type: Literal["action", "info", "warning", "danger", "text"] = "text"
+        if type in ("action", "info", "warning", "danger"):
+            self.type = type
+
+
+def find_templates(path: Path | str = ".") -> list[ShinyTemplate]:
+    path = Path(path)
+    templates: list[ShinyTemplate] = []
+
+    template_files = sorted(path.glob("**/_template.json"))
+    for tf in template_files:
+        with tf.open() as f:
+            template = json.load(f)
+
+            # "next_steps" and "follow_up" can be either a string or an array of strings
+            # or an array of dictionaries (follow_up only)
+            follow_up_raw: str | list[dict[str, str]] = template.get("follow_up", [])
+            if isinstance(follow_up_raw, str):
+                follow_up_raw = [{"text": follow_up_raw}]
+
+            follow_up = [ShinyTemplateFollowUp(**f) for f in follow_up_raw]
+
+            next_steps: str | list[str] = template.get("next_steps", [])
+            if isinstance(next_steps, str):
+                next_steps = [next_steps]
+
+            templates.append(
+                ShinyTemplate(
+                    name=template["name"],
```

Comment:
I think it's worth adding a check here as well that let's you know which template file doesn't have a `name` key